src/snowflake/snowpark/dataframe_writer.py

```diff
@@ -456,9 +456,11 @@ def save_as_table(
             )
             if column_order is None or column_order.lower() not in ("name", "index"):
                 raise ValueError("'column_order' must be either 'name' or 'index'")
-            column_names = (
-                self._dataframe.columns if column_order.lower() == "name" else None
-            )
+
+            column_names = None
+            if column_order.lower() == "name":
+                column_names = [x.name for x in self._dataframe.schema._to_attributes()]
```

Comment:
I'm trying to understand the issue
for example, for your test case which uses column name "AS".

what does `self._dataframe.columns` give you?